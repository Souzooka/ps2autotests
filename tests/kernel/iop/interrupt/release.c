#include <stdio.h>
#include <intrman.h>

void intrHandler(int value) {
}

void testNormalOperation() {
	int oldStat = 0;

	printf("normal operation:\n");

	CpuSuspendIntr(&oldStat);
	{
		ReleaseIntrHandler(IOP_IRQ_VBLANK);
		
		RegisterIntrHandler(IOP_IRQ_VBLANK, 0, (void *)&intrHandler, NULL);
		int result = ReleaseIntrHandler(IOP_IRQ_VBLANK);
		
		printf("  release after handler registered: %d\n", result);
	}
	CpuResumeIntr(oldStat);
}

void testReleaseAfterRelease() {
	int oldStat = 0;

	printf("release after release:\n");

	CpuSuspendIntr(&oldStat);
	{
		ReleaseIntrHandler(IOP_IRQ_VBLANK);
		int result = ReleaseIntrHandler(IOP_IRQ_VBLANK);
		printf("  release after handler released: %d\n", result);
	}
	CpuResumeIntr(oldStat);
}

void testInvalidParams() {
	int oldStat = 0;

	printf("invalid params:\n");

	CpuSuspendIntr(&oldStat);
	{
		int invalidIntrLineResult = ReleaseIntrHandler(~0);
		printf("  invalid interrupt line: %d\n", invalidIntrLineResult);
	}
	CpuResumeIntr(oldStat);
}

int _start(int argc, char **argv) {
	printf("-- TEST BEGIN\n");
	
	testNormalOperation();
	testReleaseAfterRelease();
	testInvalidParams();

	printf("-- TEST END\n");
	
	return 0;
}
