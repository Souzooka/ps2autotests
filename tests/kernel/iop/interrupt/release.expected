-- TEST BEGIN
normal operation:
  release after handler registered: 0
release after release:
  release after handler released: -105
invalid params:
  invalid interrupt line: -101
-- TEST END
